### PR TITLE
move QML element registration to build.rs instead of bridge

### PR DESCRIPTION
**Pull request by Be-ing**
This avoids the need to repeatedly specify the URI and version of the QML module in the bridges. QObjects that are registered as QML elements are now marked by
[cxx_qt::qobject(qml_element = "OptionalName")]

This is a precursor to generating qmldir files which are required for qmlcachegen and qmltc
(https://github.com/KDAB/cxx-qt/issues/242)

### CODE REVIEW

File: crates/cxx-qt-build/src/lib.rs [hunk 1]
```diff
@@ -241,11 +237,17 @@ fn generate_cxxqt_cpp_files(
...
         println!("cargo:rerun-if-changed={path}");
 
-        let generated_code = GeneratedCpp::new(&path)?;
+        let generated_code = match GeneratedCpp::new(&path) {
+            Ok(v) => v,
+            Err(diagnostic) => {
+                diagnostic.report();
+                std::process::exit(1);
+            }
+        };
```
Comment (Be-ing) [inline crates/cxx-qt-build/src/lib.rs:246]: Moving this error handling here avoids some duplicate code because this function is now called twice in CxxQtBuilder::build. There's nothing else useful to do with the error and the process exits here, so there's no point propagating the error upward.

File: crates/cxx-qt-gen/src/parser/qobject.rs [hunk 2]
```diff
@@ -394,17 +350,14 @@ pub mod tests {
     #[test]
     fn test_qml_metadata() {
         let item: ForeignTypeIdentAlias = parse_quote! {
-            #[cxx_qt::qobject(qml_uri = "foo.bar", qml_version = "1.0")]
+            #[cxx_qt::qobject(qml_element)]
             type MyObject = super::MyObjectRust;
         };
```
Comment (Be-ing) [inline crates/cxx-qt-gen/src/parser/qobject.rs:355]: I was thinking of having the syntax be like:
```rust
#[cxx_qt::qobject]
#[qml_element]
#[qml_singleton]
type MyObject = super::MyObjectRust;
```

but this was easier to implement and also makes sense, so :shrug:

File: examples/qml_features/rust/src/containers.rs [hunk 3]
```diff
@@ -33,7 +33,7 @@ pub mod qobject {
     }
 
     unsafe extern "RustQt" {
-        #[cxx_qt::qobject(qml_uri = "com.kdab.cxx_qt.demo", qml_version = "1.0")]
+        #[cxx_qt::qobject(qml_element)]
```
Comment (Be-ing) [inline examples/qml_features/rust/src/containers.rs:36]: This is nicer than repeating that copypasta over and over :smiley:

File: crates/cxx-qt-gen/src/parser/qobject.rs (hunk 2 above)
Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/parser/qobject.rs:355]: I think eventually the syntax will be either
```rust
#[qml_element]
#[qml_singleton]
type MyObject = super::MyObjectRust;
```
or
```rust
#[qobject]
#[qml_element]
#[qml_singleton]
type MyObject = super::MyObjectRust;
```

But what you have at the moment is fine as a stepping stone, we can sort out exact the syntax in follow ups :-)

File: examples/qml_features/rust/src/containers.rs (hunk 3 above)
Comment (ahayzen-kdab) [reply examples/qml_features/rust/src/containers.rs:36]: :tada:

File: crates/cxx-qt-gen/src/parser/qobject.rs [hunk 4]
```diff
@@ -124,63 +121,22 @@ impl ParsedQObject {
             &qobject_ty.attrs[attr_index],
             AttributeDefault::Some(|span| LitStr::new("", span)),
         )?;
-        let qml_uri = attrs_map.get(&quote::format_ident!("qml_uri"));
-        let qml_version = attrs_map.get(&quote::format_ident!("qml_version"));
-        let qml_name = attrs_map.get(&quote::format_ident!("qml_name"));
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/parser/qobject.rs:129]: i would note that we have lost `qml_name` support in this, i think this was for `QML_NAMED_ELEMENT(T)`. But this can be readded in another pull request i guess ?
```cpp
class SqlEventDatabase : public QObject
{
    Q_OBJECT
    QML_NAMED_ELEMENT(EventDatabase)

    // ...
};
```

https://doc.qt.io/qt-6/qqmlengine.html#QML_NAMED_ELEMENT

File: crates/cxx-qt-gen/src/parser/qobject.rs [hunk 5]
```diff
@@ -414,17 +367,14 @@ pub mod tests {
     #[test]
     fn test_qml_metadata_named() {
         let item: ForeignTypeIdentAlias = parse_quote! {
-            #[cxx_qt::qobject(qml_uri = "foo.bar", qml_version = "1", qml_name = "MyQmlElement")]
-            type MyNamedObject = super::MyObjectRust;
+            #[cxx_qt::qobject(qml_element = "OtherName")]
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/parser/qobject.rs:370]: ah i see you've put it within the `qml_element = "T"`

File: crates/cxx-qt-gen/src/parser/qobject.rs (hunk 4 above)
Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/parser/qobject.rs:129]: Ah i see further down this is combined into `qml_element = "T"`

File: crates/cxx-qt-gen/src/parser/qobject.rs (hunk 5 above)
Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/parser/qobject.rs:370]: Slight bit of magic but probably makes sense and avoids doing `qml_element, qml_named_element = "OtherName"` which is repetitive.

File: crates/qt-build-utils/src/lib.rs [hunk 6]
```diff
@@ -545,16 +541,13 @@ impl QtBuild {
             include_args += &format!("-I {} ", include_path.display());
         }
 
-        let mut uri_args = String::new();
-        for uri in uris {
-            uri_args += &format!("-Muri={} ", uri);
+        let mut cmd = Command::new(self.moc_executable.as_ref().unwrap());
+
+        if let Some(uri) = uri {
+            cmd.arg(&format!("-Muri={} ", uri));
```
[the remark this answers is not in the document]
Comment (Be-ing) [reply crates/qt-build-utils/src/lib.rs]: good catch, fixed

File: examples/qml_features/rust/build.rs [hunk 7]
```diff
@@ -9,19 +9,26 @@ use cxx_qt_build::CxxQtBuilder;
...
-        .file("src/multiple_qobjects.rs")
-        .file("src/nested_qobjects.rs")
-        .file("src/serialisation.rs")
-        .file("src/signals.rs")
-        .file("src/singleton.rs")
-        .file("src/properties.rs")
-        .file("src/threading.rs")
-        .file("src/types.rs")
-        .file("src/uncreatable.rs")
+        .qml_module(
```
Comment (ahayzen-kdab) [inline examples/qml_features/rust/build.rs:12]: was the plan still to move the arguments into a struct like

```rust
    CxxQtBuilder::new()
        .qml_module(QmlModule {
            uri: "com.example.qobject"
            version_major: 1
            version_minor: 0
            rust_files: ["src/cxxqt_object.rs"]
            qml_files: ["ui/foo.qml"]
        })
        .build();
```

As when we have mutliple args having named args is much nicer. And having a struct allows in the future the potential of builder pattern too.

Comment (Be-ing) [reply examples/qml_features/rust/build.rs:12]: Maybe, but then I think the struct would have to deal with lifetimes. This implementation is easy and flexible in what the user can provide with CxxQtBuilder::qml_module taking `&str` and `&[impl AsRef<Path>]` then converting those to owned values to create the struct.

File: crates/cxx-qt-build/src/lib.rs (hunk 1 above)
Comment (ahayzen-kdab) [reply crates/cxx-qt-build/src/lib.rs:246]: Sure, doesn't look like this would break anything in @LeonMatthesKDAB 's diagnostics battle :-)

**Review by ahayzen-kdab — APPROVED** [review]
LGTM :-)
